SteamKit2/SteamKit2/Networking/Steam3/UdpConnection.cs

```diff
@@ -144,11 +146,18 @@ public override void Disconnect()
             if ( netThread == null )
                 return;
 
-            // Play nicely and let the server know that we're done. Other party is expected to Ack this,
-            // so it needs to be sent sequenced.
-            SendSequenced(new UdpPacket(EUdpPacketType.Disconnect));
+            // if we think we aren't already disconnected, apply disconnecting unless we read back disconnected
+            if ( state != (int)State.Disconnected && Interlocked.Exchange(ref state, (int)State.Disconnecting) == (int)State.Disconnected )

```

Comment:
Why are you doing an `Interlocked.Exchange` here, but not doing similar thread-safe access in other areas?